src/sqlfluff/core/parser/segments/raw.py

```diff
@@ -148,3 +151,51 @@ def get_end_pos_marker(self):
     def get_start_pos_marker(self):
         """Return the pos marker at the start of this segment."""
         return self.pos_marker
+
+
+class RawSegmentFactory:
+    optional = False
+    _anti_template = None
+
+    def __init__(self, cls, classname, _template, _name, **kwargs):
+        # if classname.startswith("semicolon"):
+        #     import pdb; pdb.set_trace()
+        self.cls = cls
+        self.__name__ = classname
+        self._template = _template
+        self._name = _name
+        self._kwargs = kwargs
+
+    def __repr__(self):
+        return f"<{self.cls.__name__}: {self.__name__} {self._template} {self._name} {self._kwargs}"
+
+    def __call__(self, raw, pos_marker):
+        result = self.cls(
+            raw,
+            pos_marker,
+            _classname=self.__name__,
+            _template=self._template,
+            _name=self._name,
+            _cls=self,
+            **self._kwargs
+        )
+        return result
+
+    def match(self, segments, parse_context):
+        parse_context._factory = self
+        result = self.cls.match(segments, parse_context)
+        parse_context._factory = None
```

Comment:
Eventually, setting `_factory` might be better as a context manager. Then you can do something like:

```
with parse_context.set_factory(self):
    result = self.cls.match(segments, parse_context)
```

(ditto in `simple` below)